libs/schemas/src/Data/Schema.hs

```diff
@@ -0,0 +1,559 @@
+  deriving (Applicative) via (Const doc)
+  deriving (Profunctor, Choice) via Joker (Const doc)
+
+-- This instance is not exactly correct, distributivity does not hold
+-- in general.
+-- FUTUREWORK: introduce a NearSemiRing type class and replace the
+-- `Monoid doc` constraint with `NearSemiRing doc`.
+instance Monoid doc => Alternative (SchemaDoc doc a) where
+  empty = mempty
+  (<|>) = (<>)
+
+class HasDoc a a' doc doc' | a a' -> doc doc' where
+  doc :: Lens a a' doc doc'
+
+instance HasDoc (SchemaDoc doc a b) (SchemaDoc doc' a b) doc doc' where
+  doc = lens getDoc $ \s d -> s {getDoc = d}
+
+-- | A combined JSON encoder-decoder with associated documentation.
+--
+-- A value of type 'SchemaP d v w a b', which we will refer to as a
+-- /schema/, contains both a JSON parser and a JSON serialiser,
+-- together with documentation-like metadata, such as a JSON or
+-- Swagger schema.
+--
+-- The type variables are as follows:
+--
+--   [@d@] documentation type, usually a 'Monoid'.
+--   [@v@] type of JSON values being parsed (e.g. 'A.Value').
+--   [@w@] type of JSON values being serialised (e.g. 'A.Value').
+--   [@a@] input type
+--   [@b@] output type
+--
+-- Input and output types deserve some more explanation. We can think
+-- of a value @sch@ of type 'SchemaP d v w a b' as a kind of special
+-- "function" from @a@ to @b@, but where @a@ and @b@ might potentially
+-- live in different "languages". The parser portion of @sch@ takes a
+-- JSON-encoded value of type @a@ and produces a value of type @b@,
+-- while the serialiser portion of @sch@ takes a haskell value of type
+-- @a@ and produces a JSON-encoding of something of type @b@.
+--
+-- In terms of composability, this way of representing schemas (based
+-- on input and output types) is superior to the perhaps more natural
+-- approach of using "bidirectional functions" or isomorphisms (based
+-- on a single type parameter).
+--
+-- Although schemas cannot be composed as functions (i.e. they do not
+-- form a 'Category'), they still admit a number of important and
+-- useful instances, such as 'Profunctor' (and specifically 'Choice'),
+-- which makes it possible to use prism quite effectively to build
+-- schema values.
+--
+-- Using type variables to represent JSON types might seem like
+-- excessive generality, but it is useful to represent "intermediate"
+-- schemas arising when building complex ones. For example, a schema
+-- which is able to work with fields of a JSON object (see 'field')
+-- should not output full-blown objects, but only lists of pairs, so
+-- that they can be combined correctly via the usual 'Monoid'
+-- structure of lists when using the 'Applicative' interface of
+-- 'SchemaP d v w a b'.
+--
+-- The idea of using the profunctor structure of 'SchemaP' is taken
+-- from the [codec](https://github.com/chpatrick/codec) library.
+data SchemaP doc v w a b
+  = SchemaP
+      (SchemaDoc doc a b)
+      (SchemaIn v a b)
+      (SchemaOut w a b)
+  deriving (Functor)
```

Comment:
What does the `P` in `SchemaP` stand for?

Reply:
Profunctor